snuba/subscriptions/entity_subscription.py

```diff
@@ -0,0 +1,133 @@
+    SNQL = "snql"
+    DELEGATE = "delegate"
+
+
+class EntitySubscription(ABC):
+    def __init__(
+        self, subscription_type: SubscriptionType, data_dict: Mapping[str, Any]
+    ) -> None:
+        self.subscription_type = subscription_type
+
+    def get_entity_subscription_conditions(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Union[Expression, Condition]]:
+        return []
+
+    def to_dict(self) -> Mapping[str, Any]:
+        return {}
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, EntitySubscription):
+            return NotImplemented
+        return self.to_dict() == other.to_dict() and isinstance(other, type(self))
+
+
+class SessionsSubscription(EntitySubscription):
+    organization: int
+
+    def __init__(
+        self, subscription_type: SubscriptionType, data_dict: Mapping[str, Any]
+    ):
+        super().__init__(subscription_type, data_dict)
+        try:
+            self.organization = data_dict["organization"]
+        except KeyError:
+            raise InvalidQueryException(
+                "organization param is required for any query over sessions entity"
+            )
+
+    def get_entity_subscription_conditions(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Union[Expression, Condition]]:
+        """
+        Returns a list of extra conditions that are entity specific and required for the
+        subscription
+        """
+        conditions_to_add = []
+        if self.subscription_type == SubscriptionType.SNQL:
+            conditions_to_add += [
+                binary_condition(
+                    ConditionFunctions.EQ,
+                    Column(None, None, "org_id"),
+                    Literal(None, self.organization),
+                ),
+            ]
+        return conditions_to_add
+
+    def to_dict(self) -> Mapping[str, Any]:
+        return {"organization": self.organization}
+
+
+class EventsSubscription(EntitySubscription):
+    def get_entity_subscription_conditions(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Union[Expression, Condition]]:
+        """
+        Returns a list of extra conditions that are entity specific and required for the
+        subscription
+        """
+        if offset is None:
+            return []
+
+        conditions_to_add: List[Union[Expression, Condition]] = []
+        if self.subscription_type == SubscriptionType.SNQL:
+            conditions_to_add.append(
+                binary_condition(
+                    ConditionFunctions.LTE,
+                    FunctionCall(
+                        None,
+                        "ifNull",
+                        (Column(None, None, "offset"), Literal(None, 0)),
+                    ),
+                    Literal(None, offset),
+                )
+            )
+        elif self.subscription_type == SubscriptionType.LEGACY:
+            conditions_to_add = [[["ifnull", ["offset", 0]], "<=", offset]]
```

Comment:
Please use `ifNull` rather than `ifnull`. This list of functions will be enforced soon https://github.com/getsentry/snuba/blob/master/snuba/query/functions.py#L173